cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -87,6 +89,206 @@ Status GetListElementIndex(const ExecValue& value, T* out) {
   return Status::OK();
 }
 
+template <typename Type, typename IndexType>
+struct ListSlice {
+  using offset_type = typename Type::offset_type;
+
+  static Status Exec(KernelContext* ctx, const ExecSpan& batch, ExecResult* out) {
+    const auto opts = OptionsWrapper<ListSliceOptions>::Get(ctx);
+
+    // Invariants
+    if (opts.start < 0 || (opts.start >= opts.stop && opts.stop != -1)) {
+      // TODO: support start == stop which should give empty lists
+      return Status::Invalid("`start`(", opts.start,
+                             ") should be greater than 0 and smaller than `stop`(",
+                             opts.stop, ")");
+    }
+    if (opts.step != 1) {
+      // TODO: support step in slicing
+      return Status::NotImplemented(
+          "Setting `step` to anything other than 1 is not supported; got step=",
+          opts.step);
+    }
+    if (opts.stop == -1) {
+      // TODO: Support slicing to arbitrary end
```

Comment:
I did see the default `None` of SliceOptions then is represented as the max int64. I think this doesn't match for this case for at least two reasons:
- If user wanted fixed size output, then it's ambiguous if they literally put in the max int value, or if this is slice to end of the largest list item; `-1` makes that distinction.
- `-1` is a fairly common flag to say "use everything", from how many cpu cores to use or similar.

 > ... if you slice until after the end, does it also get filled with nulls, or do we automatically slice until the end (like how Python slicing works)?
 
 I think this would depend on the value of `return_fixed_size_list`

Reply:
That sounds good! However I can't seem to get around the conversion of the options to string in `string_builder` when using the `std::optional<int64_t>`

> /home/milesg/Projects/arrow/cpp/src/arrow/util/string_builder.h:49:10: error: no match for 'operator<<' (operand types are 'std::ostream' {aka 'std::basic_ostream<char>'} and 'const std::optional<long int>')
   49 |   stream << head;
      |   ~~~~~~~^~~~~~~

Reply:
You'll need to add overloads of GenericToString, GenericFromScalar, and GenericToScalar here: https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/function_internal.h

(Or, avoid the generic stuff and implement the FunctionOptionsType manually)

Reply:
Super, thank you! 
I found myself doing an `is_optional` / `enable_if_optional`; I left it in the same `function_internal.h` even though most are defined in `type_traits.h`, but figured you may have comments on how it was done so left it in the neighborhood. :)

Reply:
> If this is all ambiguous, can we use std::optional<int64_t> for the slice stop?

Switched to this, good idea. Although it came with some challenges as it's [not yet available in Cython 0.29](https://github.com/cython/cython/issues/3293#issuecomment-1223058101) but I like the more direct link to how it is represented in C++.